### PR TITLE
[ECO-4952] Start implementing room lifecycle spec

### PR DESCRIPTION
**Pull request by ruancomelli**
**Note: This is based on top of #60; please review that one first.**

This is based on https://github.com/ably/specification/pull/200 at `b4a495e`. It’s generated some questions, which I’ve asked on that PR.

Note that the spec has been through a few revisions since I started implementing this; I’ve tried to keep everything in sync but some older stuff might still be accidentally there.

I’ve implemented most of the specified behaviour for the `ATTACH`, `DETACH`, and `RELEASE` operations. I have not yet implemented `RETRY` since it’s quite different to those three and I wanted to get eyes on this first chunk of work; have created #51 for implementing it.

Of the operations that I have implemented, I have not implemented the following (this is accurately reflected by the `@spec…` tags in the tests):

- Lifecycle behaviour that relates to another ongoing operation (created #52)
- Lifecycle behaviour relating to “transient disconnect timeouts” (created #48)
- Lifecycle behaviour that occurs “asynchronously” outside an operation (created #50)
- CHA-RL1g2, which refers to emitting “discontinuity events” which are a concept not yet implemented; this will come in #53.

The room lifecycle manager introduced by this commit is currently a standalone object, which is not integrated with the rest of the SDK. I’ve created #47 for doing this integration work.

The `@spec…` tags are based on the on the [JS rules](https://github.com/ably/ably-js/blob/main/CONTRIBUTING.md#tests-alignment-with-the-ably-features-specification) @ `8c9ce8b`, but I camel-cased them and also decided that:

- if a test doesn't relate to a spec point, it doesn't need any markers
- there should be a way to know that a spec point is tested across multiple tests
- there should be a way of marking a spec point as implemented but not tested (so that can show up differently in reporting; important given that we’re also going to use this report as a to-do list of what still needs to be implemented)

Part of #28.

## Summary by CodeRabbit

- **New Features**
	- Introduced new error codes for improved error handling in the chat SDK.
	- Added a new section in the contribution guidelines for better test documentation related to SDK features.
	- Implemented a `RoomLifecycleManager` to manage chat room contributor states effectively.
	- Added a `SimpleClock` protocol to facilitate asynchronous task management in testing.
	- Introduced a mock implementation of `RoomLifecycleContributorChannel` for testing purposes.

- **Bug Fixes**
	- Enhanced error reporting for attachment and detachment failures within the chat SDK.

- **Tests**
	- Added comprehensive unit tests for the `RoomLifecycleManager` to ensure robust lifecycle management of chat rooms.
	- Introduced mock implementations to facilitate controlled testing scenarios.
	- Improved assertion capabilities in the testing framework for better error handling validation.

### CODE REVIEW

**Comment by ruancomelli** [comment]
This is an experiment review for experiment `mermaid_diagrams`
Run ID: `mermaid_diagrams/run_2024-09-26T09-30-04_v1-23-0-20-g8a080be57-dirty`
The benchmark review for this pull request can be found at https://github.com/sourcery-ai-experiments/ably-chat-swift/pull/1.
This pull request was cloned from `https://github.com/ably-labs/ably-chat-swift/pull/54`. _(Note: the URL is not a link to avoid triggering a notification on the original pull request.)_

<details>
<summary>Experiment configuration</summary>

```yaml
review_config:
  # User configuration for the review
  # - benchmark - use the user config from the benchmark reviews
  # - <value> - use the value directly
  user_review_config:
    enable_ai_review: true
    enable_rule_comments: false

    enable_complexity_comments: false
    enable_security_comments: false
    enable_tests_comments: false
    enable_comment_suggestions: false
    enable_functionality_review: false

    enable_pull_request_summary: false
    enable_review_guide: true

    enable_approvals: false

  ai_review_config:
    # The model responses to use for the experiment
    # - benchmark - use the model responses from the benchmark reviews
    # - llm - call the language model to generate responses
    model_responses:
      comments_model: benchmark
      comment_area_model: benchmark
      comment_validation_model: benchmark
      comment_suggestion_model: benchmark
      complexity_model: benchmark
      functionality_model: benchmark
      security_model: benchmark
      tests_model: benchmark
      pull_request_summary_model: benchmark
      review_guide_model: llm
      overall_comments_model: benchmark

# The pull request dataset to run the experiment on
pull_request_dataset:
- https://github.com/sourcery-ai/core/pull/4607
- https://github.com/sourcery-ai/core/pull/4631
- https://github.com/sourcery-ai/core/pull/4647
    # CodeRabbit examples:
- https://github.com/2lambda123/-Orange-OpenSource-oorobot/pull/15
- https://github.com/2lambda123/galaxyproject-galaxy/pull/12
- https://github.com/a0v0/avtoolz/pull/79
- https://github.com/adityask98/Hotaru/pull/10
- https://github.com/agdas/vscode/pull/2
- https://github.com/agluszak/hirschgarten/pull/2
- https://github.com/alexsnow348/insightface/pull/46
- https://github.com/alikuxac/utilities/pull/10
- https://github.com/AlphaDev87/timba-api/pull/49
- https://github.com/AngeloTadeucci/Maple2/pull/239
- https://github.com/AngeloTadeucci/Maple2.File/pull/36
- https://github.com/AngeloTadeucci/Maple2/pull/233
    # Examples where CodeRabbit does not generate diagrams
- https://github.com/baptisteArno/typebot.io/pull/1778
- https://github.com/btipling/foundations/pull/33
- https://github.com/btipling/foundations/pull/31
- https://github.com/chintu-777/jaeger/pull/1
- https://github.com/coji/remix-docs-ja/pull/55
- https://github.com/DaveMBush/SmartNgRX/pull/622
- https://github.com/DaveMBush/SmartNgRX/pull/481
- https://github.com/dkittle/party-connections/pull/6
- https://github.com/Drajad-Kusuma-Adi/onstudy-backend/pull/6
- https://github.com/imaami/libcanth/pull/2
  # Requested by Tim
- https://github.com/2lambda123/-Orange-OpenSource-oorobot/pull/15
- https://github.com/2lambda123/galaxyproject-galaxy/pull/12
- https://github.com/a0v0/avtoolz/pull/79
- https://github.com/adityask98/Hotaru/pull/10
- https://github.com/agdas/vscode/pull/2
- https://github.com/agluszak/hirschgarten/pull/2
- https://github.com/4DNucleome/PartSeg/pull/1183
- https://github.com/ably/ably-cocoa/pull/1973
- https://github.com/ably-labs/ably-chat-swift/pull/54
- https://github.com/ably-labs/ably-chat-swift/pull/35

# Questions to ask to label the review comments
review_comment_labels: []
# - label: correct
#   question: Is this comment correct?

# Benchmark reviews generated by running
#   python -m scripts.experiment benchmark <experiment_name>
benchmark_reviews:
- dataset_pull_request: https://github.com/sourcery-ai/core/pull/4607
  review_pull_request: https://github.com/sourcery-ai-experiments/core/pull/338
- dataset_pull_request: https://github.com/sourcery-ai/core/pull/4631
  review_pull_request: https://github.com/sourcery-ai-experiments/core/pull/339
- dataset_pull_request: https://github.com/sourcery-ai/core/pull/4647
  review_pull_request: https://github.com/sourcery-ai-experiments/core/pull/340
- dataset_pull_request: https://github.com/2lambda123/-Orange-OpenSource-oorobot/pull/15
  review_pull_request: https://github.com/sourcery-ai-experiments/-Orange-OpenSource-oorobot/pull/2
- dataset_pull_request: https://github.com/2lambda123/galaxyproject-galaxy/pull/12
  review_pull_request: https://github.com/sourcery-ai-experiments/galaxyproject-galaxy/pull/1
- dataset_pull_request: https://github.com/adityask98/Hotaru/pull/10
  review_pull_request: https://github.com/sourcery-ai-experiments/Hotaru/pull/2
- dataset_pull_request: https://github.com/agdas/vscode/pull/2
  review_pull_request: https://github.com/sourcery-ai-experiments/vscode/pull/3
- dataset_pull_request: https://github.com/agluszak/hirschgarten/pull/2
  review_pull_request: https://github.com/sourcery-ai-experiments/hirschgarten/pull/2
- dataset_pull_request: https://github.com/alikuxac/utilities/pull/10
  review_pull_request: https://github.com/sourcery-ai-experiments/utilities/pull/2
- dataset_pull_request: https://github.com/AlphaDev87/timba-api/pull/49
  review_pull_request: https://github.com/sourcery-ai-experiments/timba-api/pull/2
- dataset_pull_request: https://github.com/AngeloTadeucci/Maple2/pull/239
  review_pull_request: https://github.com/sourcery-ai-experiments/Maple2/pull/3
- dataset_pull_request: https://github.com/AngeloTadeucci/Maple2.File/pull/36
  review_pull_request: https://github.com/sourcery-ai-experiments/Maple2.File/pull/2
- dataset_pull_request: https://github.com/AngeloTadeucci/Maple2/pull/233
  review_pull_request: https://github.com/sourcery-ai-experiments/Maple2/pull/4
- dataset_pull_request: https://github.com/baptisteArno/typebot.io/pull/1778
  review_pull_request: https://github.com/sourcery-ai-experiments/typebot.io/pull/1
- dataset_pull_request: https://github.com/btipling/foundations/pull/33
  review_pull_request: https://github.com/sourcery-ai-experiments/foundations/pull/1
- dataset_pull_request: https://github.com/btipling/foundations/pull/31
  review_pull_request: https://github.com/sourcery-ai-experiments/foundations/pull/2
- dataset_pull_request: https://github.com/chintu-777/jaeger/pull/1
  review_pull_request: https://github.com/sourcery-ai-experiments/jaeger/pull/1
- dataset_pull_request: https://github.com/coji/remix-docs-ja/pull/55
  review_pull_request: https://github.com/sourcery-ai-experiments/remix-docs-ja/pull/1
- dataset_pull_request: https://github.com/DaveMBush/SmartNgRX/pull/622
  review_pull_request: https://github.com/sourcery-ai-experiments/SmartNgRX/pull/1
- dataset_pull_request: https://github.com/DaveMBush/SmartNgRX/pull/481
  review_pull_request: https://github.com/sourcery-ai-experiments/SmartNgRX/pull/2
- dataset_pull_request: https://github.com/dkittle/party-connections/pull/6
  review_pull_request: https://github.com/sourcery-ai-experiments/party-connections/pull/1
- dataset_pull_request: https://github.com/Drajad-Kusuma-Adi/onstudy-backend/pull/6
  review_pull_request: https://github.com/sourcery-ai-experiments/onstudy-backend/pull/1
- dataset_pull_request: https://github.com/imaami/libcanth/pull/2
  review_pull_request: https://github.com/sourcery-ai-experiments/libcanth/pull/1
- dataset_pull_request: https://github.com/4DNucleome/PartSeg/pull/1183
  review_pull_request: https://github.com/sourcery-ai-experiments/PartSeg/pull/2
- dataset_pull_request: https://github.com/ably/ably-cocoa/pull/1973
  review_pull_request: https://github.com/sourcery-ai-experiments/ably-cocoa/pull/1
- dataset_pull_request: https://github.com/ably-labs/ably-chat-swift/pull/54
  review_pull_request: https://github.com/sourcery-ai-experiments/ably-chat-swift/pull/1
- dataset_pull_request: https://github.com/ably-labs/ably-chat-swift/pull/35
  review_pull_request: https://github.com/sourcery-ai-experiments/ably-chat-swift/pull/2

```

</details>

**Comment by SourceryAI** [comment]
## Reviewer's Guide by Sourcery

This pull request implements the initial room lifecycle management functionality for the Ably Chat SDK. It introduces a `RoomLifecycleManager` class that handles the ATTACH, DETACH, and RELEASE operations as specified in the Chat SDK features specification. The implementation includes error handling, state transitions, and retry mechanisms for various scenarios.

### Sequence Diagrams

#### Room Lifecycle State Diagram

```mermaid
stateDiagram-v2
    [*] --> INITIALIZED
    INITIALIZED --> ATTACHING: ATTACH
    ATTACHING --> ATTACHED: Success
    ATTACHING --> SUSPENDED: Failure (some cases)
    ATTACHING --> FAILED: Failure (some cases)
    ATTACHED --> DETACHING: DETACH
    DETACHING --> DETACHED: Success
    DETACHING --> FAILED: Failure (some cases)
    DETACHED --> ATTACHING: ATTACH
    DETACHED --> RELEASING: RELEASE
    RELEASING --> RELEASED: Success
    RELEASED --> [*]
    SUSPENDED --> ATTACHING: ATTACH
    FAILED --> ATTACHING: ATTACH
```

#### ATTACH Operation Sequence Diagram

```mermaid
sequenceDiagram
    participant C as Client
    participant RLM as RoomLifecycleManager
    participant CC as ContributorChannels
    C->>RLM: performAttachOperation()
    RLM->>RLM: changeStatus(to: .attaching)
    loop For each contributor
        RLM->>CC: attach()
        alt Success
            CC-->>RLM: Success
        else Failure
            CC-->>RLM: Error
            RLM->>RLM: Handle error (change status, detach others)
            RLM-->>C: Throw error
        end
    end
    RLM->>RLM: changeStatus(to: .attached)
    RLM-->>C: Success
```

### File-Level Changes

| Change | Details | Files |
| ------ | ------- | ----- |
| Implement room lifecycle management | <ul><li>Create RoomLifecycleManager class to handle room lifecycle operations</li><li>Implement ATTACH operation with error handling and state transitions</li><li>Implement DETACH operation with error handling and state transitions</li><li>Implement RELEASE operation with error handling and state transitions</li><li>Add retry mechanisms for failed operations</li></ul> | `Sources/AblyChat/RoomLifecycleManager.swift` |
| Add new error codes and types for room lifecycle operations | <ul><li>Introduce new error codes for attachment and detachment failures</li><li>Create ChatError enum for detailed error information</li><li>Implement error handling for various room states and operations</li></ul> | `Sources/AblyChat/Errors.swift` |
| Create mock objects and helpers for testing | <ul><li>Implement MockRoomLifecycleContributorChannel for simulating channel behavior</li><li>Create MockSimpleClock for controlling time-based operations in tests</li><li>Add helper functions for error checking in tests</li></ul> | `Tests/AblyChatTests/Mocks/MockRoomLifecycleContributorChannel.swift`<br/>`Tests/AblyChatTests/Mocks/MockSimpleClock.swift`<br/>`Tests/AblyChatTests/Helpers/Helpers.swift` |
| Implement comprehensive unit tests for RoomLifecycleManager | <ul><li>Add tests for ATTACH operation in various scenarios</li><li>Add tests for DETACH operation in various scenarios</li><li>Add tests for RELEASE operation in various scenarios</li><li>Implement tests for error handling and state transitions</li></ul> | `Tests/AblyChatTests/RoomLifecycleManagerTests.swift` |
| Update contribution guidelines for testing and documentation | <ul><li>Add guidelines for attributing tests to specification points</li><li>Introduce @spec, @specOneOf, and @specPartial tags for test documentation</li><li>Add instructions for marking untested specification points</li></ul> | `CONTRIBUTING.md` |

---

<details>
<summary>Tips and commands</summary>

#### Interacting with Sourcery

- **Trigger a new review:** Comment `@sourcery-ai review` on the pull request.
- **Continue discussions:** Reply directly to Sourcery's review comments.
- **Generate a GitHub issue from a review comment:** Ask Sourcery to create an
  issue from a review comment by replying to it.

#### Customizing Your Experience

Access your [dashboard](https://app.sourcery.ai) to:
- Enable or disable review features such as the Sourcery-generated pull request
  summary, the reviewer's guide, and others.
- Change the review language.
- Add, remove or edit custom review instructions.
- Adjust other review settings.

#### Getting Help

- [Contact our support team](mailto:support@sourcery.ai) for questions or feedback.
- Visit our [documentation](https://docs.sourcery.ai) for detailed guides and information.
- Keep in touch with the Sourcery team by following us on [X/Twitter](https://x.com/SourceryAI), [LinkedIn](https://www.linkedin.com/company/sourceryai/) or [GitHub](https://github.com/sourcery-ai).

</details>

File: Sources/AblyChat/Errors.swift [hunk 1]
```diff
@@ -16,37 +16,163 @@ public enum ErrorCode: Int {
...
+    case roomIsReleasing = 102_102
+    case roomIsReleased = 102_103
+
     /// The ``ARTErrorInfo.statusCode`` that should be returned for this error.
     internal var statusCode: Int {
-        // TODO: These are currently a guess, revisit in https://github.com/ably-labs/ably-chat-swift/issues/32
+        // TODO: These are currently a guess, revisit once outstanding spec question re status codes is answered (https://github.com/ably/specification/pull/200#discussion_r1755222945), and also revisit in https://github.com/ably-labs/ably-chat-swift/issues/32
         switch self {
-        case .inconsistentRoomOptions:
+        case .inconsistentRoomOptions,
```
Comment (SourceryAI) [inline Sources/AblyChat/Errors.swift:39]: **suggestion:** Simplify the switch statement in the statusCode property

Consider grouping the cases that return the same value to reduce repetition and improve readability. For example, you could use a single case with a pattern that matches all relevant error codes.

File: Sources/AblyChat/Errors.swift [hunk 2]
```diff
@@ -16,37 +16,163 @@ public enum ErrorCode: Int {
...
+                .typingDetachmentFailed
+            }
+        case .roomInFailedState:
+            .roomInFailedState
+        case .roomIsReleasing:
+            .roomIsReleasing
+        case .roomIsReleased:
+            .roomIsReleased
+        }
+    }
```
Comment (SourceryAI) [inline Sources/AblyChat/Errors.swift:111]: **suggestion:** Extract repetitive code in localizedDescription into a separate method

The switch statements for attachment and detachment failures contain similar code. Consider extracting this logic into a separate method to reduce duplication and improve maintainability.

**Review by SourceryAI — COMMENTED** [review]
Hey @ruancomelli - I've reviewed your changes - here's some feedback:

**Overall Comments**:
- Ensure that all TODOs and references to issues are tracked and resolved to avoid technical debt.
- Revisit error handling and status codes once specification questions are resolved to maintain consistency.
- Ensure comprehensive test coverage for all parts of the specification, especially those marked as `@specPartial` or `@specUntested`.

<details>
<summary>Here's what I looked at during the review</summary>

- 🟡 **General issues**: 2 issues found
- 🟢 **Documentation**: all looks good
</details>

***

[LangSmith trace](https://smith.langchain.com/o/7d8518ad-6dd3-5c7b-96d2-757769296335/projects/p/88560225-5072-48bc-967c-dc54da53f265/r/e7cd514b-ceea-4291-9fda-759a6b7f8ce0?poll=true)

<sub>
Help me be more useful! Please click 👍 or 👎 on each comment to tell me if it was helpful.
</sub>

**Pull request closed without merging**